crypto/rsa/rsa_ossl.c

```diff
@@ -624,11 +628,11 @@ static int rsa_ossl_mod_exp(BIGNUM *r0, const BIGNUM *I, RSA *rsa, BN_CTX *ctx)
         BN_with_flags(p, rsa->p, BN_FLG_CONSTTIME);
         BN_with_flags(q, rsa->q, BN_FLG_CONSTTIME);
 
-        if (rsa->flags & RSA_FLAG_CACHE_PRIVATE) {
-            if (!BN_MONT_CTX_set_locked
-                (&rsa->_method_mod_p, rsa->lock, p, ctx)
-                || !BN_MONT_CTX_set_locked(&rsa->_method_mod_q,
-                                           rsa->lock, q, ctx)) {
+        if (smooth || (rsa->flags & RSA_FLAG_CACHE_PRIVATE)) {
```

Comment:
This means that if the `smooth` codepath runs on a key that `RSA_FLAG_CACHE_PRIVATE` was cleared, we change behavior and the RSA now remembers bits of itself, which I believe is user-visible if the caller changes the key after use. Perhaps `smooth` needs to require `RSA_FLAG_CACHE_PRIVATE`, or perhaps the non-`RSA_FLAG_CACHE_PRIVATE` case needs to create local `BN_MONT_CTX` and pass that in.

This complexity, incidentally, is a big part of why I filed #5158.

(We took away `RSA_FLAG_CACHE_PRIVATE` and always act as if it is on.)

Reply:
Yeah, I was anticipating this question... I mean I know that it's kind of an edge and still am thinking about it...